### DISCUSSION TITLE
How to use W3CTraceContextPropagator to read the traceID from header

### DISCUSSION THREAD

**Discussion by arunprakashpj** [open-telemetry/opentelemetry-java-instrumentation#6968] [Q&A]
As of now we are sending a trace parent  ( ‘traceparent: 00-xxxxxxxxxxxxxxxxxxxxxx-yyyyyyyyyyyy-01’) from cloud events, and we expect open telemetry to print the console logs with trace IDs and span Ids based on the header.

How to make use of the automatic instrumentation to print trace IDs and span IDs in console logs of the spring boot app ? The trace ID has to be automatically fetched from trace parent initially.

**Comment by Ipuvi** [comment]
If using automatic instrumentation then, the OTel Java agent injects information about the current span into each logging
event's MDC copy:

- `trace_id` - the current trace id
  (same as `Span.current().getSpanContext().getTraceId()`);
- `span_id` - the current span id
  (same as `Span.current().getSpanContext().getSpanId()`);
- `trace_flags` - the current trace flags, formatted according to W3C traceflags format
  (same as `Span.current().getSpanContext().getTraceFlags().asHex()`).

These can be included in log statements produced by the logging library
by specifying them in the pattern/format, for Spring Boot configuration which uses logback, you can add MDC to log lines by overriding only the `logging.pattern.level`:

```properties
logging.pattern.level = trace_id=%mdc{trace_id} span_id=%mdc{span_id} trace_flags=%mdc{trace_flags} %5p
```

**Comment by santoshkashyap** [comment]
@arunprakashpj , I'm not sure if you are able to get it done automatically ? Even I have a similar situation of processing CloudEvents in a spring boot application. As of now, I think there is no auto-instrumentation support for CloudEvents in OpenTelemetry Java Agent (See, [supported-libraries](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/supported-libraries.md)). For e.g. as of today based on the `traceparent` attribute in the CloudEvent, OpenTelemetry(SDK/JavaAgent) does not automatically continue the trace by generating a new span for CloudEvent processing.  

For my app, I tried to handle producer and consumer spans manually based on the current experimental spec: [CloudEvents OTel](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/cloudevents/). Because, the spans are manually created using OpenTelemetry API, the logging fields `trace_id`,  `span_id` are not present in the application logs. I think you need to use the MDC Apis to manually add these fields and remove them after processing:

```
// parse traceparent to extract required traceId, spanId etc.
final SpanContext activeSpanContext = SpanContext.createFromRemoteParent(traceId, spanId, traceFlag,
						TraceState.getDefault());
final Span activeSpan = Span.wrap(activeSpanContext);
activeContext = Context.current().with(activeSpan);


// create a span with "activeContext"
final SpanBuilder currentSpanBuilder = tracer.spanBuilder(OPERATION_NAME)
.setSpanKind(SpanKind.CONSUMER)
.setParent(activeContext);

// create span 
final Span currentSpan = currentSpanBuilder.startSpan();

try(Scope scope = currentSpan.makeCurrent()) {
// store trace_id in MDC for logging. similarly for other fields.
//MDC.put("trace_id", Span.current().getSpanContext().getTraceId())

// logic to process cloud event. 

}
finally{
// clear trace_id after processing.
MDC.remove("trace_id")
}
```
